Add extended providers support (#292)
filecoin-project/index-provider: README.md

* [`engine/example_test.go`](engine/example_test.go)

#### Publishing ads with extended providers

[Extended providers](https://github.com/filecoin-project/storetheindex/blob/main/doc/ingest.md#extendedprovider) 
field allows for specification of provider families, in cases where a provider operates multiple PeerIDs, perhaps 
with different transport protocols between them, but over the same database of content. 

> Such ads can be composed manually or using a conveniece builder `ExtendedProvidersAdBuilder`.

Reviewer: ```suggestion
Such ads can be composed manually or using a convenience builder `ExtendedProvidersAdBuilder`.
```

```

  adv, err := ep.NewExtendedProviderAdBuilder(providerID, priv, addrs). // the main ad's providerID, private key and addresses
    WithContextID(contextID). // optional context id
    WithMetadata(metadata). // optional metadata
    WithOverride(override). // override flag, false by default

>     WithExtendedProviders(extendedProviders). // one or more extended providers to be included into the ad, represnted by ExtendedProviderInfo struct

Reviewer: ```suggestion
    WithExtendedProviders(extendedProviders). // one or more extended providers to be included in the ad, represented by ExtendedProviderInfo struct
```

>     WithLastAdID(lastAdId). // cid of the last publisghed ad, which is false by default

Reviewer: ```suggestion
    WithLastAdID(lastAdId). // cid of the last published ad, which is false by default
```

To create one use

```
publisher := NewExtendedProvidersPublisher(
  "12D3KooWK2QCtvVYsCTCdGA3bsey1nzHdd4NrvFVsoP5Ze4k1NDV", // peer id of the main provider - the one that will be specified in the ad body
  "CAESQP300ou/EsEnReN1BTvtjedhAjdAJGEydytYNwt6sAiM50lPfT64jA5P/RDf9o9IY0r+VtHdGgJYxyNAD0KJg=", // private key of the main provider
  []ma.Multiaddrs{"/ip4/167.176.1.98/tpc/3090", "/ip4/167.176.1.98/tpc/3091"}, // multiaddresses of the main provider
  engine, // pointer to the underlying engine for publishing ads
  NewExtendedProviderInfo( // one or multiple extended provider infos
    "12D3KooWK2QCtvVYsCTCdGA3bsey1nzHdd4NrvFVsoP5Ze4k1NDV", // peer id of the extended provider
    []ma.Multiaddrs{"/ip4/174.176.1.98/tpc/3090", "/ip4/184.176.1.98/tpc/3091"} // multiaddresses where the extended provider is available at
    []byte("some metadata"), // optional metadata that is included into the extended provider ad
    "ABCSQP300ou/EsEnReN1BTvtjedhAjdAJGEydytYNwt6sAiM50lPfT64jA5P/RDf9o9IY0r+VtHdGgJYxyNAD0KJg=" // prviate key of the extended provider that is going to be used to sign its records in the ads
  ), ....
)
```

To publish an ad

```

> publisher.Publish(

Author: ok, will modify it to be compilable
